docs/contributing/debugging.md

```diff
@@ -0,0 +1,33 @@
+### Debugging
+
+Debugging java agent can be a challenging task since some instrumentation
+code is directly inlined into target classes and debugger is
+usually not attached early enough to activate breakpoints
+in agent initialization code (`premain`/`agentmain`, `agent-tooling` artifact etc.).
+
+#### Template methods and agent initialization
+
+Breakpoints do not work in template methods, because their code is directly inlined
+by ByteBuddy into the target class. It is good to keep these methods as small as possible.
+The template methods are annotated with:
+
+```java
+@net.bytebuddy.asm.Advice.OnMethodEnter
+@net.bytebuddy.asm.Advice.OnMethodExit
+```
+
+The best approach to debug template methods and agent initialization is to use the following statements:
+
+```java
+System.out.println()
+Thread.dumpStack()
+```
+
+#### Enable debugger
+
+The following example shows remote debugger configuration. The breakpoints
```

Comment:
IIUC, this requires remote debugging from this repo since it has the code, but not a users if they are using the agent. Since this is CONTRIBUTING.md that's totally fine :) With the linked issue I'm hoping we can provide that jar so it's easier to debug for a user though - add compileOnly dependency on obfuscated jar and use normal IntelliJ local debug, it's worked ok for me in the past.